dds/idl/typeobject_generator.cpp

```diff
@@ -6,499 +6,1740 @@
+    }
+    if (this->value & OpenDDS::XTypes::IS_OPTIONAL) {
+      out << " | XTypes::IS_OPTIONAL";
+    }
+    if (this->value & OpenDDS::XTypes::IS_MUST_UNDERSTAND) {
+      out << " | XTypes::IS_MUST_UNDERSTAND";
+    }
+    if (this->value & OpenDDS::XTypes::IS_KEY) {
+      out << " | XTypes::IS_KEY";
     }
-    return;
+    if (this->value & OpenDDS::XTypes::IS_DEFAULT) {
+      out << " | XTypes::IS_DEFAULT";
+    }
+    return out;
   }
+};
 
-  if (fld_cls & CL_ARRAY) {
-    // TODO [anonymous]: Is an anonymous array always plain?
-    AST_Array* array = AST_Array::narrow_from_decl(type);
+struct CollectionElementFlagPrinter : public MemberFlagPrinter {
+  CollectionElementFlagPrinter(const OpenDDS::XTypes::CollectionElementFlag a_value) : MemberFlagPrinter(a_value) {}
+};
 
-    ACE_CDR::ULong max_bound = 0;
+struct StructMemberFlagPrinter : public MemberFlagPrinter {
+  StructMemberFlagPrinter(const OpenDDS::XTypes::StructMemberFlag a_value) : MemberFlagPrinter(a_value) {}
+};
 
-    for (ACE_CDR::ULong dim = 0; dim != array->n_dims(); ++dim) {
-      max_bound = std::max(max_bound, array->dims()[dim]->ev()->u.ulval);
-    }
+struct UnionMemberFlagPrinter : public MemberFlagPrinter {
+  UnionMemberFlagPrinter(const OpenDDS::XTypes::StructMemberFlag a_value) : MemberFlagPrinter(a_value) {}
+};
 
-    if (max_bound > 255) {
-      be_global->impl_ <<
-        "XTypes::TypeIdentifier::makePlainArray(";
-      call_get_minimal_type_identifier(array->base_type());
-      be_global->impl_ << ", XTypes::LBoundSeq()";
-      for (ACE_CDR::ULong dim = 0; dim != array->n_dims(); ++dim) {
-        be_global->impl_ << ".append(XTypes::LBound(" << array->dims()[dim]->ev()->u.ulval << "))";
-      }
-      be_global->impl_ << ")";
-    } else {
-      be_global->impl_ <<
-        "XTypes::TypeIdentifier::makePlainArray(";
-      call_get_minimal_type_identifier(array->base_type());
-      be_global->impl_ << ", XTypes::SBoundSeq()";
-      for (ACE_CDR::ULong dim = 0; dim != array->n_dims(); ++dim) {
-        be_global->impl_ << ".append(XTypes::SBound(" << array->dims()[dim]->ev()->u.ulval << "))";
-      }
-      be_global->impl_ << ")";
+struct UnionDiscriminatorFlagPrinter : public MemberFlagPrinter {
+  UnionDiscriminatorFlagPrinter(const OpenDDS::XTypes::StructMemberFlag a_value) : MemberFlagPrinter(a_value) {}
+};
+
+struct EnumeratedLiteralFlagPrinter : public MemberFlagPrinter {
+  EnumeratedLiteralFlagPrinter(const OpenDDS::XTypes::StructMemberFlag a_value) : MemberFlagPrinter(a_value) {}
+};
+
+struct AliasMemberFlagPrinter : public MemberFlagPrinter {
+  AliasMemberFlagPrinter(const OpenDDS::XTypes::AliasMemberFlag a_value) : MemberFlagPrinter(a_value) {}
+};
+
+struct TypeFlagPrinter : public ValuePrinter<OpenDDS::XTypes::TypeFlag> {
+  TypeFlagPrinter(const OpenDDS::XTypes::EquivalenceKind a_value) : ValuePrinter(a_value) {}
+
+  std::ostream& print_on(std::ostream& out) const
+  {
+    out << "0";
```

Comment:
This is a little kludgy, but I'm OK with it.  Another approach would be to keep track of when the first output happens and only add `|` after that.